### PR TITLE
Auth md processor

### PR DESCRIPTION
**Pull request by jboeuf**
Plug-in server-side auth metadata processing.
Right now the processor is a global but it should really on the server_credentials object (or as another parameter in the port).

### CODE REVIEW

File: src/core/security/server_auth_filter.c [hunk 1]
```diff
@@ -31,20 +31,149 @@
...
+    grpc_metadata_batch_filter(&calld->md_op->data.metadata, remove_consumed_md,
+                               elem);
+    GPR_ASSERT(calld->call_auth_context != NULL);
+    GRPC_AUTH_CONTEXT_UNREF(*calld->call_auth_context,
+                            "releasing old context.");
+    *calld->call_auth_context =
+        GRPC_AUTH_CONTEXT_REF(result, "refing new context.");
+    calld->on_done_recv->cb(calld->on_done_recv->cb_arg, success);
+  } else {
+    grpc_transport_stream_op_add_cancellation(
```
Comment (jboeuf) [inline src/core/security/server_auth_filter.c]: This is the error case that does not behave properly. I'm not quite sure what to do here. If I call `calld->on_done_recv->cb(...)` after the cancellation, I get an assert failure.

File: test/core/end2end/tests/request_response_with_payload_and_call_creds.c [hunk 2]
```diff
@@ -340,31 +437,128 @@ static void request_response_with_payload_and_call_creds(
...
+  GPR_ASSERT(GRPC_CALL_OK == grpc_call_start_batch(c, ops, op - ops, tag(1)));
+
+  cq_expect_completion(cqv, tag(1), 1);
+  cq_verify(cqv);
+
+  /* XXX Should be GRPC_STATUS_UNAUTHENTICATED but it looks like there is a bug
+     (probably in the server_auth_context.c code) where this error on the server
+     does not get to the client. The current error code we are getting is
+     GRPC_STATUS_INTERNAL. */
+  GPR_ASSERT(status != GRPC_STATUS_OK);
```
Comment (jboeuf) [inline test/core/end2end/tests/request_response_with_payload_and_call_creds.c]: Here I would expect to get a `GRPC_STATUS_UNAUTHENTICATED` which is what (I think) I'm setting on the server.

**Comment by jboeuf** [comment]
Sorry if there's a lot of noise on this PR as I had to move a few things around. However, please have a look at the changes in server_auth_filter.c.

**Comment by jboeuf** [comment]
ping (the word of the day :)).

File: src/core/security/server_auth_filter.c (hunk 1 above)
Comment (ctiller) [inline src/core/security/server_auth_filter.c]: I have a day or two's work to get this working. Can this move forward without it (with a bug filed with what needs to change when I get to it)?

Comment (jboeuf) [inline src/core/security/server_auth_filter.c]: Thanks Craig. Let's wait for your fix and not introduce this new API surface before we're sure that everything works well (I can still file an issue though).

Comment (ctiller) [inline src/core/security/server_auth_filter.c]: That'd be good

**Comment by ctiller** [comment]
Sent https://github.com/jboeuf/grpc/pull/2 to you against this to sort out the status problems.

**Comment by jboeuf** [comment]
go home CLA. you're drunk...

**Comment by jboeuf** [comment]
Closed in favor of #2777 that has a better API.

**Pull request closed without merging**